libfwupdplugin/fu-common.c

```diff
@@ -197,6 +198,88 @@ fu_common_check_full_disk_encryption(GError **error)
 	return TRUE;
 }
 
+/**
+ * fu_common_get_olson_timezone_id:
+ * @error: (nullable): optional return location for an error
+ *
+ * Gets the system Olson timezone ID, as used in the CLDR and ICU specifications.
+ *
+ * Returns: timezone string, e.g. `Europe/London` or %NULL on error
+ *
+ * Since: 1.9.7
+ **/
+gchar *
+fu_common_get_olson_timezone_id(GError **error)
+{
+	g_autofree gchar *sysconfdir = fu_path_from_kind(FU_PATH_KIND_SYSCONFDIR);
+	g_autofree gchar *fn_lz = fu_path_from_kind(FU_PATH_KIND_LOCALTIME);
+	g_autofree gchar *fn_tz = g_build_filename(sysconfdir, "timezone", NULL);
+	g_autoptr(GFile) file_lz = g_file_new_for_path(fn_lz);
+	g_autoptr(GFile) file_tz = g_file_new_for_path(fn_tz);
+
+	/* old Ubuntu, or no systemd */
+	g_debug("looking for timezone file %s", fn_tz);
+	if (g_file_query_exists(file_tz, NULL)) {
+		gsize bufsz = 0;
+		g_autofree gchar *buf = NULL;
+		g_auto(GStrv) lines = NULL;
+		if (!g_file_load_contents(file_tz, NULL, &buf, &bufsz, NULL, error))
+			return NULL;
+		lines = fu_strsplit(buf, bufsz, "\n", -1);
+		for (guint i = 0; lines[i] != NULL; i++) {
+			if (lines[i][0] == '\0')
+				continue;
+			if (g_str_has_prefix(lines[i], "#"))
+				continue;
+			return g_strdup(lines[i]);
+		}
+	}
+
+	/* Red Hat and *BSD use the last two sections of the symlink target */
+	g_debug("looking for timezone file %s", fn_lz);
+	if (g_file_query_file_type(file_lz, G_FILE_QUERY_INFO_NOFOLLOW_SYMLINKS, NULL) ==
+	    G_FILE_TYPE_SYMBOLIC_LINK) {
+		const gchar *target = NULL;
+		g_autoptr(GFileInfo) info = NULL;
+
+		info = g_file_query_info(file_lz,
+					 G_FILE_ATTRIBUTE_STANDARD_SYMLINK_TARGET,
+					 G_FILE_QUERY_INFO_NOFOLLOW_SYMLINKS,
+					 NULL,
+					 error);
+		if (info == NULL)
+			return NULL;
+		target = g_file_info_get_symlink_target(info);
+		if (target != NULL) {
+			g_auto(GStrv) sections = g_strsplit(target, "/", -1);
+			guint sections_len = g_strv_length(sections);
+			if (sections_len < 2) {
+				g_set_error(error,
+					    G_IO_ERROR,
+					    G_IO_ERROR_INVALID_FILENAME,
+					    "invalid symlink target: %s",
+					    target);
+				return NULL;
+			}
+			return g_strdup_printf("%s/%s",
+					       sections[sections_len - 2],
+					       sections[sections_len - 1]);
+		}
+	}
+
+#ifdef _WIN32
+	/* there does not appear to be any API for this */
+	return g_strdup("America/New_York");
```

Comment:
for a more predictable result shouldn't you just return `NULL` for `_WIN32` too?

Reply:
That means that it doesn't work with fwupdtool.exe -- I'm using it in the logitech bulkcontroller plugin which should work there too. I've asked Logi for how they do it in the Win32 updater they have.

Reply:
Try using https://learn.microsoft.com/en-us/windows/win32/api/timezoneapi/nf-timezoneapi-gettimezoneinformation and then use this table to convert it: https://github.com/uddhav/tzid-from-geo/blob/master/olson_to_windows.json

Reply:
Or using something like this specifically with Windows perhaps:
https://github.com/mattjohnsonpint/TimeZoneConverter

Reply:
Good idea, I guess I can use the JSON file mapping in reverse.

Reply:
Done.